File: karma/src/main/java/com/sns/karma/util/JwtUtil.java
package com.sns.karma.util;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.security.Keys;

import java.nio.charset.StandardCharsets;
import java.security.Key;
import java.util.Date;

public class JwtUtil {
    public static String generateToken(String username, String key, long duration){
        Claims claims = Jwts.claims();
        claims.put(key, username);
        Long currentTimeMillis = System.currentTimeMillis();
        return Jwts.builder()
                .setClaims(claims)
                .setIssuedAt(new Date(currentTimeMillis))               // 발행일자
                .setExpiration(new Date(currentTimeMillis+duration))    // 만료일자
                .signWith(generateKey(key) ,SignatureAlgorithm.HS256)   // 키 & 암호화 알고리즘
                .compact();                                             // toString
    }
    private static Key generateKey(String key){
        byte[] keyBytes = key.getBytes(StandardCharsets.UTF_8);
        return Keys.hmacShaKeyFor(keyBytes);
    }
}

File: backend/src/main/kotlin/com/karma/chat/util/JwtUtil.kt
package com.karma.chat.util

import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.security.Key
import java.util.*

@Component
class JwtUtil(
    @Value("\${jwt.secret}")
    private val jwtSecret: String,
    @Value("\${jwt.expire}")
    private val expire: Int,
    @Value("\${jwt.issuer}")
    private val issuer: String
) {

    val key: Key by lazy {
        Keys.hmacShaKeyFor(Decoders.BASE64.decode(jwtSecret))
    }

    fun generateToken(subject: String) = Jwts.builder()
        .signWith(key, SignatureAlgorithm.HS512)
        .setSubject(subject)
        .setIssuer(issuer)
        .setIssuedAt(Date())
        .setExpiration(Date(System.currentTimeMillis() * 60 * expire))
        .compact()!!

    fun extractSubject(token: String): String? = Jwts.parserBuilder()
        .setSigningKey(key)
        .build()
        .parseClaimsJws(token)
        .body
        .subject
}